models/balancer/arbitrum/balancer_v2_arbitrum_bpt_supply.sql

```diff
@@ -0,0 +1,514 @@
+            ] AS amounts_out
+        FROM ( --Start subquery
+            SELECT 
+                x.contract_address
+                , x.evt_tx_hash
+                , x.evt_index
+                , x.evt_block_time
+                , x.evt_block_number
+                , x.poolId AS poolId
+                , sp.pool_token AS pool_token
+                , x.deltas
+                , x.deltas[ARRAY_POSITION(x.tokens, sp.pool_token)] AS delta_pool_token
+                , y.lp_token_burned
+                , y.lp_token_minted
+            FROM {{ source('balancer_v2_arbitrum', 'Vault_evt_PoolBalanceChanged') }} x --#########################
+            INNER JOIN (SELECT poolId, tokens, pool_token FROM sub_pools) sp
+                ON sp.poolId = x.poolId
+            INNER JOIN (
+                SELECT 
+                    evt_tx_hash
+                    , contract_address
+                    , IF(
+                        sum(CAST(value AS int256)) FILTER(
+                            WHERE "from" = 0x0000000000000000000000000000000000000000
+                        ) IS NULL
+                        , int256 '0'
+                        , sum(CAST(value AS int256)) FILTER(
+                            WHERE "from" = 0x0000000000000000000000000000000000000000
+                        )
+                    ) AS lp_token_minted
+                    , IF(
+                        sum(CAST(value AS int256)) FILTER(
+                            WHERE to = 0x0000000000000000000000000000000000000000
+                        ) IS NULL
+                        , int256 '0'
+                        , sum(CAST(value AS int256)) FILTER(
+                            WHERE to = 0x0000000000000000000000000000000000000000
+                        )
+                    ) AS lp_token_burned
+                FROM {{ source('erc20_arbitrum', 'evt_transfer') }}  --#########################
+                WHERE evt_block_number >= (SELECT min(evt_block_number) FROM sub_pools) 
+                    AND (to = 0x0000000000000000000000000000000000000000 
+                        OR "from" = 0x0000000000000000000000000000000000000000
+                    )
+                    AND contract_address in (SELECT pool_token FROM sub_pools)
+                GROUP BY 1, 2
+            ) y
+            ON y.evt_tx_hash = x.evt_tx_hash
+                AND y.contract_address = sp.pool_token
+        ) x -- End subquery
+        -- Join transactions to get additional data
+        INNER JOIN (
+            SELECT
+                block_number
+                , hash AS tx_hash
+                , index AS tx_index
+                , "from" AS tx_from
+                , "to" AS tx_to
+            FROM {{ source('arbitrum', 'transactions') }} --#########################
+            WHERE block_number >= (SELECT min(evt_block_number) FROM sub_pools)
+        ) l
+        ON l.block_number = x.evt_block_number
+            AND l.tx_hash = x.evt_tx_hash
+    ) -- End CTE
```

Comment:
not a must, but we intro'd a new macro to grab columns from `<blockchain>.transactions` base table:
https://github.com/duneanalytics/spellbook/blob/main/macros/models/_sector/add_tx_columns.sql

we grab the same three columns in new `dex.trades_beta` spell that we're redesigning. thought you may find it helpful here.

Reply:
Would it be used like so:
```
        -- Join transactions to get additional data
        INNER JOIN (
            add_tx_columns(
                model_cte = 'transactions'
                , blockchain = 'arbitrum'
                , columns = [block_number, hash AS tx_hash, index AS tx_index, "from" AS tx_from, "to" AS tx_to] 
            )
            WHERE block_number >= (SELECT min(evt_block_number) FROM sub_pools)
        ) l
        ON l.block_number = x.evt_block_number
            AND l.tx_hash = x.evt_tx_hash
    ) -- End CTE
```

Reply:
not quite. you would remove lines 149-161, end the CTE after the `--end subquery` comment. then remove any column selections (like line 54) in original CTE. start a new CTE after, something like:

```
with join_exit_call_data_enhanced as (
  {{
      add_tx_columns(
          model_cte = 'join_exit_call_data' --CTE one above
          , blockchain = 'arbitrum'
          , columns = ['index'] --noticed you only pull `index` column
      )
  }}
)
```

then you could use `join_exit_call_data_enhanced` downstream instead

Reply:
might be worth a quick try, then run `dbt compile` and see the output sql looks as expected compared to now (with an additional CTE to add the column from base table). if it gets too complicated, don't worry about it, but could be a nice clean, consistent approach

Reply:
no problem, you can leave as-is.

the goal with macros is to centralize repeatable code, for easier maintenance moving forward & also less work for future contributions. i think there are various use cases:
- in `dex.trades` lineage, there are many uniswap forks. we had a ton of models writing the same code, just flipping project names and such. now, we centralize the logic into one place, then all models simply call the macro. any future changes are in one spot.
- joins from decoded tables to base tables, such as the macro suggested here, to ensure consistency across how we join. there are times where joins are written differently across models. another case, while probably rare/less likely, is if the base tables were to change how they are partitioned, it might make sense to tweak the join conditions, which would allow us to edit one spot rather than hundreds, if not more
- one project across various blockchains, one macro for the same logic especially if evm-based, then just the parameters passed in change

Reply:
Gotcha thanks for elaborating on that!